File: learn/layer2-mechanism/linked-list/linked_list.c
/*
 * Layer 2 · 메커니즘 — 링크드 리스트 (C: 포인터 체인이 명시적으로 드러남)
 *
 * Python 버전이 감추는 것을 C 는 그대로 보여준다:
 *   - 노드는 malloc 으로 힙 아무 데나 할당된다(연속 아님)
 *   - '연결'은 다음 노드의 주소(Node*)를 필드에 저장하는 것일 뿐
 *   - 삭제는 앞 노드의 next 를 뒤 노드 주소로 바꾸고, 중간 노드를 free
 *
 * 실행: gcc -O2 -o /tmp/ll linked_list.c && /tmp/ll   (통과 시 'OK')
 */
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

typedef struct Node {
    int val;
    struct Node *next;   /* 다음 노드의 '주소' — 이게 포인터 체인의 실체 */
} Node;

typedef struct {
    Node *head;
    int   size;
} List;

static void list_init(List *l) { l->head = NULL; l->size = 0; }

static void push_front(List *l, int val) {            /* O(1) */
    Node *n = malloc(sizeof(Node));
    n->val = val;
    n->next = l->head;
    l->head = n;
    l->size++;
}

static void push_back(List *l, int val) {             /* O(n) — tail 미보유 */
    Node *n = malloc(sizeof(Node));
    n->val = val; n->next = NULL;
    if (!l->head) { l->head = n; }
    else {
        Node *cur = l->head;
        while (cur->next) cur = cur->next;            /* 끝까지 걸어감 */
        cur->next = n;
    }
    l->size++;
}

static int find(List *l, int val) {                   /* O(n) — 순차 탐색 */
    int idx = 0;
    for (Node *cur = l->head; cur; cur = cur->next, idx++)
        if (cur->val == val) return idx;
    return -1;
}

static int delete_val(List *l, int val) {             /* O(n) 탐색 + O(1) 잇기 */
    Node *prev = NULL, *cur = l->head;
    while (cur) {
        if (cur->val == val) {
            if (!prev) l->head = cur->next;
            else       prev->next = cur->next;        /* 중간 노드 분리 */
            free(cur);                                /* 메모리 해제(Python 은 GC 가 함) */
            l->size--;
            return 1;
        }
        prev = cur; cur = cur->next;
    }
    return 0;
}

static void free_all(List *l) {
    Node *cur = l->head;
    while (cur) { Node *nx = cur->next; free(cur); cur = nx; }
    l->head = NULL; l->size = 0;
}

int main(void) {
    List l; list_init(&l);
    push_back(&l, 1); push_back(&l, 2); push_back(&l, 3);
    push_front(&l, 0);                 /* [0,1,2,3] */
    assert(l.size == 4);
    assert(find(&l, 2) == 2);
    assert(find(&l, 99) == -1);
    assert(delete_val(&l, 2) == 1);    /* [0,1,3] */
    assert(delete_val(&l, 0) == 1);    /* head 삭제 [1,3] */
    assert(l.head->val == 1 && l.head->next->val == 3);
    assert(delete_val(&l, 99) == 0);
    free_all(&l);
    printf("OK\n");
    return 0;
}

File: learn/layer2-mechanism/linked-list/linked_list.py
"""
Layer 2 · 메커니즘 — 링크드 리스트 (내장 컬렉션 없이 바닥부터)
Notion: (concept page URL 은 이 폴더 README.md 참조)

이 파일이 답하는 질문:
  "어레이(연속 메모리)와 링크드리스트(포인터 체인)의 삽입/삭제/탐색 비용이
   왜 다른지 메모리 구조로 설명할 수 있는가"

포인트: 노드는 메모리 어디에 흩어져 있어도 되고, 서로를 '참조(next)'로 잇는다.
- push_front : O(1)   — head 참조만 갈아끼움. 뒤 원소 이동 없음(어레이는 O(n) shift)
- push_back  : O(n)   — tail 을 안 들면 끝까지 걸어가야 함
- find       : O(n)   — 인덱스 임의접근 불가. 순차 탐색뿐(어레이는 O(1) 임의접근)
- delete     : O(n)   — 노드 자체를 찾는 데 O(n), 잇기 자체는 O(1)

실행: python3 linked_list.py   (self-test, 통과 시 'OK')
"""


class Node:
    __slots__ = ("val", "next")

    def __init__(self, val, nxt=None):
        self.val = val
        self.next = nxt


class LinkedList:
    def __init__(self):
        self.head = None
        self._size = 0

    def push_front(self, val):          # O(1)
        self.head = Node(val, self.head)
        self._size += 1

    def push_back(self, val):           # O(n) — tail 미보유
        node = Node(val)
        if self.head is None:
            self.head = node
        else:
            cur = self.head
            while cur.next is not None:  # 끝까지 걸어감
                cur = cur.next
            cur.next = node
        self._size += 1

    def find(self, val):                # O(n) — 순차 탐색
        cur, idx = self.head, 0
        while cur is not None:
            if cur.val == val:
                return idx
            cur, idx = cur.next, idx + 1
        return -1

    def delete(self, val):              # O(n) 탐색 + O(1) 잇기
        prev, cur = None, self.head
        while cur is not None:
            if cur.val == val:
                if prev is None:
                    self.head = cur.next
                else:
                    prev.next = cur.next   # 앞 노드가 뒤 노드를 직접 참조 → 중간 노드 분리
                self._size -= 1
                return True
            prev, cur = cur, cur.next
        return False

    def to_list(self):
        out, cur = [], self.head
        while cur is not None:
            out.append(cur.val)
            cur = cur.next
        return out

    def __len__(self):
        return self._size


def _selftest():
    ll = LinkedList()
    ll.push_back(1); ll.push_back(2); ll.push_back(3)
    ll.push_front(0)
    assert ll.to_list() == [0, 1, 2, 3], ll.to_list()
    assert len(ll) == 4
    assert ll.find(2) == 2
    assert ll.find(99) == -1
    assert ll.delete(2) is True
    assert ll.to_list() == [0, 1, 3]
    assert ll.delete(0) is True          # head 삭제
    assert ll.to_list() == [1, 3]
    assert ll.delete(99) is False
    print("OK")


if __name__ == "__main__":
    _selftest()
